nengo_gui/components/sim_control.py

```diff
@@ -73,14 +77,56 @@ def control(self, t):
                 self.rate += (1 - decay) * rate
                 self.skipped = 1
 
+                # compute current proportion of full speed
+                self.rate_proportion = 1.0 - ((self.rate * self.delay_time) /
+                                              self.actual_model_dt)
+
         self.last_tick = now
 
+        # if we have a resired proportion, use it to control delay_time
+        if self.target_scale is not None:
+            s = self.target_scale
+            if s <=0:
+                target_delay = 0.5
+            elif self.rate == 0:
+                target_delay = self.delay_time
+            elif s >= 0.99:
+                target_delay = 0
+            else:
+                target_delay = (1-s)/s * (self.actual_model_dt /
+                                          (self.rate-self.delay_time))
+
+            if target_delay > 0.5:
+                target_delay = 0.5
+
+            self.delay_time = target_delay
+
+        # if we have a desired rate, do a simple P-controller to get there
+        if self.target_rate is not None:
+            rate_error = self.rate - self.target_rate
+            delta = rate_error * 0.0000002
+            self.delay_time += delta
+
+        if self.delay_time < 0:
+            self.delay_time = 0
+        if self.delay_time > 0.5:
+            self.delay_time = 0.5
+
+        if self.delay_time > 0:
+            self.busy_sleep(self.delay_time)
+
         # Sleeps to prevent the simulation from advancing
         # while the simulation is paused
         while self.paused and self.page.sim is not None:
             time.sleep(0.01)
             self.last_tick = None
 
+    def busy_sleep(self, delay_time):
```

Comment:
(my plan for something to try is to use time.sleep(), but accurately measure how much time it actually slept for, and use that to adjust how much time I'll sleep next time.)